.github/workflows/src/arm-auto-signoff-preview.js

```diff
@@ -0,0 +1,170 @@
+    owner = owner || inputs.owner;
+    repo = repo || inputs.repo;
+    issue_number = issue_number || inputs.issue_number;
+    head_sha = head_sha || inputs.head_sha;
+  }
+
+  return await getLabelActionImpl({
+    owner,
+    repo,
+    issue_number,
+    head_sha,
+    github,
+    core,
+  });
+}
+
+/**
+ * @param {Object} params
+ * @param {string} params.owner
+ * @param {string} params.repo
+ * @param {number} params.issue_number
+ * @param {string} params.head_sha
+ * @param {(import("@octokit/core").Octokit & import("@octokit/plugin-rest-endpoint-methods/dist-types/types.js").Api & { paginate: import("@octokit/plugin-paginate-rest").PaginateInterface; })} params.github
+ * @param {typeof import("@actions/core")} params.core
+ * @returns {Promise<LabelAction>}
+ */
+export async function getLabelActionImpl({
+  owner,
+  repo,
+  issue_number,
+  head_sha,
+  github,
+  core,
+}) {
+  if (!(await incrementalChangesToExistingResourceProvider(core))) {
+    return LabelAction.Remove;
+  }
+
+  // TODO: Try to extract labels from context (when available) to avoid unnecessary API call
+  const labels = (
+    await github.rest.issues.listLabelsOnIssue({
+      owner: owner,
+      repo: repo,
+      issue_number: issue_number,
+    })
+  ).data.map((label) => label.name);
+
+  // TODO: Also require label "ARMBestPracticesAcknowledgement"
+  const allLabelsMatch =
+    labels.includes("ARMReview") &&
+    !labels.includes("NotReadyForARMReview") &&
+    (!labels.includes("SuppressionReviewRequired") ||
+      labels.includes("Suppression-Approved"));
+
+  if (!allLabelsMatch) {
+    return LabelAction.Remove;
+  }
+
+  const checkRuns = (
+    await github.rest.checks.listForRef({
+      owner: owner,
+      repo: repo,
+      ref: head_sha,
+    })
+  ).data.check_runs;
+
+  const swaggerLintDiffs = checkRuns.filter(
+    (run) => run.name === "Swagger LintDiff",
+  );
+
+  if (swaggerLintDiffs.length > 1) {
+    throw new Error(
+      `Unexpected number of checks named 'Swagger LintDiff': ${swaggerLintDiffs.length}`,
+    );
+  }
+
+  const swaggerLintDiff =
+    swaggerLintDiffs.length === 1 ? swaggerLintDiffs[0] : undefined;
+
+  if (swaggerLintDiff && swaggerLintDiff.status === "completed") {
+    return swaggerLintDiff.conclusion === "success"
+      ? LabelAction.Add
+      : LabelAction.Remove;
+  } else {
+    // No-op if check is missing or not completed, to prevent frequent remove/add label as checks re-run
+    return LabelAction.None;
+  }
+}
+
+/**
+ * @param {import('github-script').AsyncFunctionArguments['core']} core
+ * @returns {Promise<boolean>} True if PR contains changes to existing RPs, and no new RPs
+ */
+async function incrementalChangesToExistingResourceProvider(core) {
+  core.info("incrementalChangesToExistingResourceProvider()");
+
+  const changedSwaggerFiles = await getChangedSwaggerFiles(
+    core,
+    "HEAD^",
+    "HEAD",
+    "",
+  );
+  const changedRmFiles = changedSwaggerFiles.filter((f) =>
```

Comment:
These are the type of helpers whether it be in JS or PS or both that we need to share as much as possible because it is a source of inconsistency and bugs across the various tools.

Reply:
Specifically which helper?  You mean filtering from "all changed swagger files" to "changed swagger files under resource-manager"?

I already extracted some helpers into `changed-files.js`, are you suggesting more?

I agree 100% in sharing this code as much as possible, definitely within a language (a single helper used by all code in that language), and potentially calling from JS into PWSH and vice-versa.

Reply:
Yes we have similar filtering in the PS scripts and other tools and I worry we might decide what is resource-manager slightly different in each place.